flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/AbstractBuffersUsageGauge.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.metrics;
+
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.metrics.Gauge;
+import org.apache.flink.runtime.io.network.partition.consumer.SingleInputGate;
+
+/**
+ * Abstract gauge implementation for calculate the buffer usage percent.
+ */
+public abstract class AbstractBuffersUsageGauge implements Gauge<Float> {
+
+	protected SingleInputGate[] inputGates;
+
+	@VisibleForTesting
+	public abstract int calculateUsedBuffers(SingleInputGate inputGate);
+
+	@VisibleForTesting
+	public abstract int calculateBufferPoolSize(SingleInputGate inputGate);
+
+	AbstractBuffersUsageGauge(SingleInputGate[] inputGates) {
+		this.inputGates = inputGates;
+	}
+
+	@Override
+	public Float getValue() {
+		int usedBuffers = 0;
+		int bufferPoolSize = 0;
+
+		for (SingleInputGate inputGate : inputGates) {
```

Comment:
I ever had a concern with the case of multiple input gates. Each `SingleInputGate` corresponds to a different upstream `JobVertex`. If we calculate the value for all the gates, it seems difficult to distinguish the specific usage for individual gate for performance tuning.

Another option is we calculate the value separately for each gate, and take the maximum value as the final value if we agree the max usages should bring more concerns for performance tuning. I am not quite sure of it. WDYT?@pnowojski

Reply:
My opinion here: although we calculated the maximum value we seems can't not know the reported value comes from which inputGate? Unless we report the individual gate metric with gate specific on metric name, but it seems too detailed with too much metric data.

Reply:
Yes, I only pointed out a different idea and have not thought through it yet. 
The current way keeps the previous behavior and I could accept it atm if no better solution.

Reply:
Agree with you that mixing serval inputGate usage will make the performance tuning harder, looking forward to better solution , I will also keep an eye on this, too :)

Reply:
I haven't thought about this. I would keep it as it is in order for this to be merged before 1.9 feature freeze and this has already grown substantially from its initial form. There can be some follow up work in the future.

Reply:
agree